engine/overworld/weather.asm

```diff
@@ -0,0 +1,778 @@
+; bc = max screen x/y coords
+	ld a, [wCurrentWeather]
+	and a
+	ret z
+	ld a, [wCurrentWeather]
+	dec a ; rain
+	jr z, .rain
+.snow_loop
+	push bc
+	call ScanForEmptyOAM
+	pop bc
+	ret c
+	ld a, c
+	call RandomRange
+	ld [hli], a
+	ld a, b
+	call RandomRange
+	ld [hli], a
+	ld a, SNOWFLAKE_TILE
+	ld [hli], a
+	ld a, 6 ; pallete 6
+	ld [hld], a
+	dec hl
+	dec hl
+	ldh a, [hUsedWeatherSpriteIndex]
+	cp l
+	jr nc, .snow_loop
+	ld a, l
+	ldh [hUsedWeatherSpriteIndex], a
+	jr .snow_loop
+
+.rain
+	push bc
+	call ScanForEmptyOAM
+	pop bc
+	ret c
+	ld a, c
+	call RandomRange
+	ld [hli], a
+	ld a, b
+	call RandomRange
+	ld [hli], a
+	call Random
+	cp 20 percent ; 20 percent splashes
+	; a = carry ? RAINSPLASH_TILE : RAINDROP_TILE
+	assert RAINDROP_TILE + 1 == RAINSPLASH_TILE
+	ccf
+	sbc a
+	add RAINSPLASH_TILE
+.got_tile
+	ld [hli], a
+	ld a, 6 ; pallete 6
+	ld [hld], a
+	dec hl
+	dec hl
+	ldh a, [hUsedWeatherSpriteIndex]
+	cp l
+	jr nc, .rain
+	ld a, l
+	ldh [hUsedWeatherSpriteIndex], a
+	jr .rain
+
+DoOverworldSnow:
+	ld a, [wLoadedObjPal6]
+	inc a
+	jr z, .continue
+	farcall LoadWeatherPal
+.continue
+	call ScanForEmptyOAM
+	call nc, SpawnSnowFlake
+	call ScanForEmptyOAM
+	call nc, SpawnSnowFlake
+; fallthrough
+DoSnowFall:
+	ld de, wShadowOAM
+	ld hl, wShadowOAM
+	ld b, NUM_SPRITE_OAM_STRUCTS
+.loop
+	ld hl, SPRITEOAMSTRUCT_YCOORD
+	ld a, [hl]
+	cp OAM_YCOORD_HIDDEN
+	jr z, .next
+	ld hl, SPRITEOAMSTRUCT_TILE_ID
+	add hl, de
+	ld a, [hli]
+	cp SNOWFLAKE_TILE
+	jr nz, .next
+	ld a, [hl]
+	cp 6 ; pallete 6
+	jr nz, .next
+
+	call Random
+	cp 1 percent
+	jr nc, .ok
+	call Random
+	cp 10 percent
+	jr c, .despawn
+.ok
+
+	ld a, [wPlayerStepVectorY]
+	add a
+	ld c, a
+	ld hl, SPRITEOAMSTRUCT_YCOORD
+	add hl, de
+	ld a, [hl]
+	sub c
+	ld c, a
+	call GetDropSpeedModifier
+	add c
+	add 2
+	ld hl, SPRITEOAMSTRUCT_YCOORD
+	add hl, de
+	cp OAM_YCOORD_HIDDEN
+	ld [hl], a
+	jr nc, .despawn
+
+	ld a, [wPlayerStepVectorX]
+	add a
+	ld c, a
+	call Random
+	and 1
+	ld a, c
+	jr nz, .no_add_1
+	inc a
+.no_add_1
+	ld c, a
+	ld hl, SPRITEOAMSTRUCT_XCOORD
+	add hl, de
+	ld a, [hl]
+	sub c
+	inc a
+	ld hl, SPRITEOAMSTRUCT_XCOORD
+	add hl, de
+	sub 1 ; no-optimize a++|a-- (need to set carry)
+	ld [hl], a
+	jr c, .despawn
+.next
+	ld hl, SPRITEOAMSTRUCT_LENGTH
+	add hl, de
+	ld d, h
+	ld e, l
+	dec b
+	jr nz, .loop
+	ret
+
+.despawn
+	ld hl, SPRITEOAMSTRUCT_YCOORD
+	add hl, de
+	ld a, OAM_YCOORD_HIDDEN
+	ld [hli], a
+	xor a
+	ld [hli], a
+	ld [hli], a
+	ld [hl], a
+	jr .next
+
+DoOverworldRain:
+	ld a, [wLoadedObjPal6]
+	cp PAL_OW_RAIN
+	jr z, .continue
+	farcall LoadWeatherPal
+.continue
+	ld a, [wCurrentWeather]
+	cp OW_WEATHER_THUNDERSTORM
+	jr nz, .no_lightning
+	call Random
+	cp 1 percent
+	jr nc, .no_lightning
+	call Random
+	cp 50 percent
+	call c, Lightning ; Do you get the reference?
+.no_lightning
+	call ScanForEmptyOAM
+	call nc, SpawnRainDrop
+	call ScanForEmptyOAM
+	call nc, SpawnRainDrop
+	call ScanForEmptyOAM
+	call nc, SpawnRainDrop
```

Comment:
`rept` is great for clarifying that yes, this should be done N times (and making N a visible number).
```suggestion
rept 3 ; spawn three raindrops at once
	call ScanForEmptyOAM
	call nc, SpawnRainDrop
endr
```